File: ast.py
class Number:
    def __init__(self, value):
        self.value = value

class String:
    def __init__(self, value):
        self.value = value

class Identifier:
    def __init__(self, name):
        self.name = name

class BinOp:
    def __init__(self, left, op, right):
        self.left = left
        self.op = op
        self.right = right

class FuncCall:
    def __init__(self, name, args):
        self.name = name
        self.args = args

class VarDecl:
    def __init__(self, var_type, name, value):
        self.var_type = var_type
        self.name = name
        self.value = value

class Assign:
    def __init__(self, name, value):
        self.name = name
        self.value = value

class Program:
    def __init__(self, stmts):
        self.stmts = stmts

File: rust/src/ast.rs
#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    Int,
    Float,
    Char,
    Pointer(Box<Type>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinOpType {
    Add,
    Sub,
    Mul,
    Div,
    Less,
    Greater,
    LessEq,
    GreaterEq,
    Eq,
    NotEq,
}

#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOpType {
    AddressOf,
    Deref,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expression {
    Number(i32),
    String(String),
    Identifier(String),
    BinOp(Box<Expression>, BinOpType, Box<Expression>),
    UnaryOp(UnaryOpType, Box<Expression>),
    FuncCall(String, Vec<Expression>),
    ArrayAccess(String, Box<Expression>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Statement {
    VarDecl(Type, String, Option<Expression>),
    ArrayDecl(Type, String, i32),
    Assign(Expression, Expression),
    Expression(Expression),
    If(Expression, Box<Vec<Statement>>, Option<Box<Vec<Statement>>>),
    While(Expression, Box<Vec<Statement>>),
    For(
        Box<Statement>, // Init (Assign or VarDecl)
        Expression,     // Condition
        Box<Statement>, // Update (Assign)
        Box<Vec<Statement>>, // Body
    ),
    Return(Expression),
    FunctionDef(String, Vec<(Type, String)>, Type, Box<Vec<Statement>>),
    Import(String),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Program {
    pub statements: Vec<Statement>,
}
